### PR TITLE
custom resource strategy: grace value added to the delete option

### LINKED ISSUE #56567: Grace period on DeleteOptions has no effect on CRs
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:

We use a CR with a custom finalizer to get support for graceful deletes. While the crd apiserver seems to consider the deleteGracePeriod to some degree, it seems that it is not updating `metadata.deletionGracePeriodSeconds`. Therefore our finalizer has no way to differentiate between a forceful delete request `--grace-period 0` or a graceful delete request.

A follow up issue seems to be that `--grace-period` is not sent by kubectl if it talks to a cr. Therefore I tried it with a DeleteOptions object on a plain curl call too.

**What you expected to happen**:

I would expect that I can implement graceful delete based on a finalizer on my CR, in combination with sending `gracePeriodSecons` to the apiserver. This allows me to distinguish between a force delete and a graceful delete.

**How to reproduce it (as minimally and precisely as possible)**:

Here the kubectl log:

```bash
$ delete vms  testvm --grace-period 20 -v9
Request Body: {"orphanDependents":false}
I1129 12:13:33.298274     728 round_trippers.go:417] curl -k -v -XDELETE  -H "Accept: application/json" -H "Content-Type: application/json" -H "User-Agent: .kubectl/v1.8.4 (linux/amd64) kubernetes/9befc2b" https://192.168.121.164:6443/apis/kubevirt.io/v1alpha1/namespaces/default/virtualmachines/testvm
I1129 12:13:33.312421     728 round_trippers.go:436] DELETE https://192.168.121.164:6443/apis/kubevirt.io/v1alpha1/namespaces/default/virtualmachines/testvm 202 Accepted in 14 milliseconds
I1129 12:13:33.312486     728 round_trippers.go:442] Response Headers:
I1129 12:13:33.312542     728 round_trippers.go:445]     Content-Type: application/json
I1129 12:13:33.312549     728 round_trippers.go:445]     Content-Length: 1198
I1129 12:13:33.312554     728 round_trippers.go:445]     Date: Wed, 29 Nov 2017 11:13:33 GMT
I1129 12:13:33.312621     728 request.go:836] Response Body: {"apiVersion":"kubevirt.io/v1alpha1","kind":"VirtualMachine","metadata":{"clusterName":"","creationTimestamp":"2017-11-29T11:07:54Z","deletionGracePeriodSeconds":0, [...]}
```
As you can see, no `gracePerodSeconds` is sent to the apiserver and the resulting delete response has `deletionGracePeriodSecons` set to 0.

And here the plain curl call:

```bash
$ curl -k -v -XDELETE  -H "Content-Type: application/json" -H "Accept: application/json" -H "User-Agent: .kubectl/v1.8.4 (linux/amd64) kubernetes/9befc2b" http://192.168.200.2:8184/apis/kubevirt.io/v1alpha1/namespaces/default/virtualmachines/testvm --data '{"orphanDependents":false, "gracePeriodSeconds": 20}'
{"metadata":{"name":"testvm","namespace":"default","selfLink":"/apis/kubevirt.io/v1alpha1/namespaces/default/virtualmachines/testvm","uid":"b7e81365-d4f6-11e7-8ec0-525400ea43c6","resourceVersion":"54319","creationTimestamp":"2017-11-29T11:16:17Z","deletionTimestamp":"2017-11-29T11:16:24Z","deletionGracePeriodSeconds":0, [...]}
```
This time I made sure by hand, that `gracePeriodSecons` is sent to the apiserver, but the resulting object still show 0 for `deletionGracePeriodSeconds`.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.4", GitCommit:"9befc2b8928a9426501d3bf62f72849d5cbcd5a3", GitTreeState:"clean", BuildDate:"2017-11-20T05:28:34Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.4", GitCommit:"9befc2b8928a9426501d3bf62f72849d5cbcd5a3", GitTreeState:"clean", BuildDate:"2017-11-20T05:17:43Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by poonai**
**What this PR does / why we need it**:
Grace value will updated if user request for grace period in the request body.
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #56567

**Special notes for your reviewer**:
do we have any default grace value(if finalizer set in the custom resource)?
**Release note**:

```release-note
grace period will updated on deleting custom resource
```

### CODE REVIEW

**Comment by thockin** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/60744)

**Comment by hzxuzhonghu** [comment]
@sch00lb0y run ./hack/update-bazel.sh

**Comment by sch00lb0y** [comment]
@hzxuzhonghu I updated. but test is not passing.
```
I0306 05:48:27.626] --- FAIL: TestDelete (0.17s)
I0306 05:48:27.626] 	resttest.go:765: unexpected, object foo1 should have been deleted immediately
I0306 05:48:27.626] 	resttest.go:769: expected status of delete, got nil
I0306 05:48:27.626] 	resttest.go:777: unexpected error: <nil>
I0306 05:48:27.627] 	resttest.go:801: unexpected error: StorageError: key exists, Code: 2, Key: /registry/noxus/test/foo1, ResourceVersion: 0, AdditionalErrorMsg: 
I0306 05:48:27.628] 	resttest.go:810: unexpected error: Operation cannot be fulfilled on Noxu.mygroup.example.com "foo1": the UID in the precondition (UID0000) does not match the UID in record (). The object might have been deleted and then recreated
I0306 05:48:27.628] 	resttest.go:815: expected status of delete, got nil
I0306 05:48:27.628] 	resttest.go:823: unexpected error: <nil>
```
I tried by curl. cr is deleted as expected and returning success status

**Comment by sch00lb0y** [comment]
```go
func (t *Tester) testDeleteNoGraceful(obj runtime.Object, createFn CreateFunc, getFn GetFunc, isNotFoundFn IsErrorFunc) {
	ctx := t.TestContext()

	foo := obj.DeepCopyObject()
	t.setObjectMeta(foo, t.namer(2))
	if err := createFn(ctx, foo); err != nil {
		t.Errorf("unexpected error: %v", err)
	}
	objectMeta := t.getObjectMetaOrFail(foo)
	obj, wasDeleted, err := t.storage.(rest.GracefulDeleter).Delete(ctx, objectMeta.GetName(), metav1.NewDeleteOptions(10))
	if err != nil {
		t.Errorf("unexpected error: %v", err)
	}
	if !wasDeleted {
		t.Errorf("unexpected, object %s should have been deleted immediately", objectMeta.GetName())
	}
	if !t.returnDeletedObject {
		if status, ok := obj.(*metav1.Status); !ok {
			t.Errorf("expected status of delete, got %v", status)
		} else if status.Status != metav1.StatusSuccess {
			t.Errorf("expected success, got: %v", status.Status)
		}
	}

	_, err = getFn(ctx, foo)
	if err == nil || !isNotFoundFn(err) {
		t.Errorf("unexpected error: %v", err)
	}
}
```
This test is not passing,
I have small quesiton.
cr is expected to not to delete if it has graceperiod right?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd.go [hunk 1]
```diff
@@ -91,7 +91,8 @@ func newREST(resource schema.GroupResource, listKind schema.GroupVersionKind, st
 		UpdateStrategy: strategy,
 		DeleteStrategy: strategy,
 
-		TableConvertor: tableConvertor,
+		TableConvertor:      tableConvertor,
+		ReturnDeletedObject: true,
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd.go]: where is this used?

Comment (sch00lb0y) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd.go]: In `testDeleteGracefulImmediate`, returned delete object is used to check the timestamps
```go
objectMeta = t.getObjectMetaOrFail(out)
	// the second delete shouldn't update the object, so the objectMeta.GetDeletionGracePeriodSeconds() should eqaul to the value set in the first delete.
	if objectMeta.GetDeletionTimestamp() == nil || objectMeta.GetDeletionGracePeriodSeconds() == nil || *objectMeta.GetDeletionGracePeriodSeconds() != 0 {
		t.Errorf("unexpected deleted meta: %#v", objectMeta)
	}
	if generation >= objectMeta.GetGeneration() {
		t.Error("Generation wasn't bumped when deletion timestamp was set")
	}
```

Comment (sch00lb0y) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd.go]: I think I should remove here and skip the test. 
`CheckGracefulDelete` is used only in pod so I took inspiration from there

**Comment by rmohr** [comment]
> cr is expected to not to delete if it has graceperiod right?

If there is no opt-in on the crd level, you probably need to proceed with a normal delete, if it is not blocked by a finalizer. @sttts was there a decision which direction you want to go?

**Pull request closed without merging**